File: Euler/CH2.cpp
#include <iostream>

using namespace std;

/**
 * @brief 
 * 
 * @tparam T (generic type) 
 * @param n (limit to sum up to n) 
 * @return T (return sum of even fibonacci numbers up to n)
 */
template <typename T>
T evenFibonacciNumbers(T n) {
    T sum = 0;
    T a = 0;
    T b = 1;
    while (b < n) {
        if (b % 2 == 0) {
            sum += b;
        }
        T temp = a;
        a = b;
        b += temp;
    }
    return sum;
}


int main(int argc, char const *argv[])
{
    cout << evenFibonacciNumbers(4000000) << endl;
    return 0;
}

File: src/main/java/Euler/CH2.java
package Euler;

public class CH2 {

    /*
     * Even fibonacci numbers
     */
    public static int evenFibonacciNumbers(int limit) {
        int a = 0;
        int b = 1;
        int sum = 0;
        while (a < limit) {
            if (a % 2 == 0) {
                sum += a;
            }
            int swap = b;
            b += a;
            a = swap;
        }
        return sum;
    }

    public static void main(String[] args) {
        long result = evenFibonacciNumbers(4000000);
        System.out.println("Answer: " + result);
    }
}
